### ISSUE TITLE
AspectFill image cropping different on iOS and UWP (Xamarin Forms)

### ISSUE THREAD

**Issue by dominik-weber** [luberda-molinet/FFImageLoading#266]
Thank you for this wonderful library! It has solved so many issues I've had with the default Xamarin Forms Image view. I found a small bug though:

When displaying an image with AspectFill the displayed portion of the image is cropped from the center of the image on iOS but from the top left corner on UWP.
My image is wider than it needs to be and is included in the xaml like so:
`<ffimage:CachedImage Source="{Binding BackgroundImage}" Aspect="AspectFill" />`

This is the source image:
![image](https://cloud.githubusercontent.com/assets/1771927/16539994/d0b37d90-4054-11e6-8872-3c57d466a014.png)

It looks like this on iOS (which is what I want):
![image](https://cloud.githubusercontent.com/assets/1771927/16539979/4ffd8f7e-4054-11e6-9181-5bb81fc8beae.png)
However it looks like this on UWP:
![image](https://cloud.githubusercontent.com/assets/1771927/16539980/67034902-4054-11e6-82f3-e9ea2891f216.png)

On a related note, did you consider including a property which could be used to define where to crop the image, eg. center, topleft, left,...?

**Comment by daniel-luberda** [comment]
Hi @dominik-weber,

It's because of the difference how Windows handles image aspect ratios. Read more here: http://stackoverflow.com/questions/13535022/windows-8-image-uniformfill-centered

Xamarin.Forms `Image` has the same issue. If you have any idea how could we solve it, please let me know. 

> On a related note, did you consider including a property which could be used to define where to crop the image, eg. center, topleft, left,...?

Sure we could add that (I already added some properties for transformations instead using constructors), but it's a low priority for now. If you have something - feel free to collaborate.

**Comment by dominik-weber** [comment]
I think this could be fixed by setting the HorizontalAlignment and VerticalAlignment properties of the Image to Center. See this pull request: https://github.com/xamarin/Xamarin.Forms/pull/190

In the screenshot below I tried setting these properties on the middle item (Contoso Tax). Notice how the red borders expand beyond the visible area of the image. If I set HorizontalAlignment to Stretch the image expands outside the visible area to the right. If I set it to Center it expands equally on both sides.

![image](https://cloud.githubusercontent.com/assets/1771927/16565716/6d2a74ba-420f-11e6-895a-019da6c22ede.png)

**Comment by daniel-luberda** [comment]
Thanks for a help! It's fixed, prerelease nuget coming soon.

**Comment by dominik-weber** [comment]
Nice, thanks! :+1:

**Comment by daniel-luberda** [comment]
Prerelease NuGet package containing the fix released. Could you test it?

**Comment by dominik-weber** [comment]
works perfectly, thanks! :)

**Comment by dominik-weber** [comment]
I'm afraid this issue isn't completely fixed... I did some more testing and figured the following:

- Fill: works fine
- AspectFill: works fine
- AspectFit: doesn't work

In this screenshot I've added an image whose layout is illustrated by the red rectangle, with Aspect set to AspectFit. On iOS the image is centered however on UWP it is aligned to the left.
![image](https://cloud.githubusercontent.com/assets/1771927/17512210/0e26d664-5e27-11e6-85e4-cd76d4c87baa.png)

I've created a sample app to test all different Aspect values and XForms Horizontal/VerticalOptions values and found that when setting the UWP Control.HorizontalAlignment and Control.VerticalAlignment to Center it works fine in all scenarios. So I would suggest removing the alignment modifications from the UpdateAspect method and just setting it during control initialization:

                var control = new Image()
                {
                    Stretch = GetStretch(Xamarin.Forms.Aspect.AspectFill),
                    HorizontalAlignment = HorizontalAlignment.Center,
                    VerticalAlignment = VerticalAlignment.Center
                };
                control.ImageOpened += OnImageOpened;
                SetNativeControl(control);

**Comment by dominik-weber** [comment]
@daniel-luberda have you seen the comment i've added?

**Comment by daniel-luberda** [comment]
@dominik-weber Hi, I've missed that, sorry. Sure, we can change it to have a unified behavior.

**Comment by dominik-weber** [comment]
just fyi, here is the project I've used to test the different scenarios on UWP. Uncomment all lines from UWP/ImageExtRenderer to see the differences.

[FFImageCropTest.zip](https://github.com/luberda-molinet/FFImageLoading/files/488652/FFImageCropTest.zip)

**Comment by dominik-weber** [comment]
Thank you for the quick implementation @daniel-luberda ! Let me know when it's available on nuget and I will run some tests.

**Comment by daniel-luberda** [comment]
2.1.8-pre-155 released. Could you test it?:)

**Comment by dominik-weber** [comment]
Unfortunately no, does not work :(
I've checked the sources of XForms, you can see that the Alignment properties are overwritten here: https://github.com/adamped/Xamarin.Forms/blob/6a23187f26aeea3e33d2f3c9a77172f062c7a519/Xamarin.Forms.Platform.WinRT/VisualElementRenderer.cs#L290
So we need to set them **after** calling SetNativeControl:

            if (Control == null)
            {
                Image control = new Image()
                {
                    Stretch = GetStretch(Xamarin.Forms.Aspect.AspectFill),
                };
                control.ImageOpened += OnImageOpened;
                SetNativeControl(control);

                // must set alignment properties after SetNativeControl
                Control.HorizontalAlignment = HorizontalAlignment.Center;
                Control.VerticalAlignment = VerticalAlignment.Center;
            }

**Comment by daniel-luberda** [comment]
Thank you for your help and debugging. I'll push a commit with fixes now.

**Comment by daniel-luberda** [comment]
2.1.8-pre-158 could you test it? There were also some changes how transformations are handled on Windows, look here: https://github.com/luberda-molinet/FFImageLoading/commit/f90bd976a43e9a15bb88b68cbb603a30182bac56

Is transformations performance better now?

**Comment by dominik-weber** [comment]
works 👍 thank you very much! :)
I only use transformations to show a very small user image as a circle, didn't notice any difference...

**Comment by dominik-weber** [comment]
Hmm... I ran some more tests regarding transformations on UWP:

                for (int i = 0; i < 1000; i++)
                {
                    var image = new ImageExt();
                    image.Source = "https://blog.optimizely.de/files/2014/01/test.jpg";

                    for (int j = 0; j < 1000; j++)
                        image.Transformations.Add(new BlurredTransformation());

                    contentLayout.Children.Add(image);
                }

So add 1000 times the same image, applying 1000 blurred transforms each.
However I did not notice any difference:

![image](https://cloud.githubusercontent.com/assets/1771927/18799547/72ce1814-81d8-11e6-9111-166a3b83190c.png)
![image](https://cloud.githubusercontent.com/assets/1771927/18799551/76d7ef20-81d8-11e6-841a-6ca7cb823e90.png)

Lots of GCing though ;)

**Comment by daniel-luberda** [comment]
Hi @dominik-weber,

Thanks for such a detailed tests, those are really useful as I don't work a lot with UWP currently. Actually, with that code, only one transformation operation was performed. All other images were loaded from cache. You would need to use a different image for every image loading operation or disable memory cache for that images. With recent changes, I just added a lock which allows only one transformation operation to perform at a time to reduce abnormal cpu/memory usage. That's why I was wondering if the transformation performance has changed.

Thanks,
Daniel

**Comment by dominik-weber** [comment]
Alright, I wasn't sure about that (although I was impressed by the performance ;) )
I have modified the test to load individual images:

                contentLayout.Children.Clear();

                var random = new Random();

                for (int i = 0; i < 200; i++)
                {
                    var image = new ImageExt();
                    image.Source = "https://blog.optimizely.de/files/2014/01/test.jpg" + "?r=" + random.Next();

                    for (int j = 0; j < 5; j++)
                        image.Transformations.Add(new BlurredTransformation());

                    contentLayout.Children.Add(image);
                }

And ran this test with 50x50 (50 images with 50 transforms each) and a much more realistic configuration of 200 images with 5 transforms each.

![image](https://cloud.githubusercontent.com/assets/1771927/18808202/01e5269a-825c-11e6-82ec-eebec10443bf.png)
![image](https://cloud.githubusercontent.com/assets/1771927/18808207/09414f4a-825c-11e6-8139-ccea40b6c07a.png)
![image](https://cloud.githubusercontent.com/assets/1771927/18808206/05fc2efe-825c-11e6-837d-5576b0eb3f69.png)
![image](https://cloud.githubusercontent.com/assets/1771927/18808209/0ba8df8c-825c-11e6-9ad2-e60c23acdbf8.png)

You can see that with the new version the cpu usage and memory consumption are much smoother. However particularly with the 200x5 example the difference in memory consumption is negligible and the duration it takes for the test to complete is a bit higher. The UI was responsive during all tests, at least on my desktop. Did not run the test on a mobile phone.
So from running this test I'd say the results are inconclusive, we have to observe more real world scenarios before we can tell if the change was a success or not.
Do you have a real use case that inspired you to implement the change?

**Comment by daniel-luberda** [comment]
@dominik-weber Impressive profiling! :) Thanks a lot for that. 

It was changed mainly because of mobile phones. There were cases where doing a lot of image processing at once caused application crash (because of abnormal cpu/memory usage). The lock eliminates that. 

On the other hand, UWP is used also on desktop apps. Is there any way to determine if UWP app is run on desktop? If yes, we could lock only on mobile phones, preserving performance on a desktop apps.

**Comment by dominik-weber** [comment]
You could check this with the following code:
`if (Windows.System.Profile.AnalyticsInfo.VersionInfo.DeviceFamily == "Windows.Desktop")`
However this seems a bit "hacked"... also there are still desktop pcs with very little RAM.

https://msdn.microsoft.com/de-de/library/windows.system.profile.analyticsversioninfo.devicefamily.aspx
> The values of this property will change over time as new device families are released. You should not depend on the value of this property, it is intended for analytics use only. For example, you could prioritize marketing activities based on data that more of your customers are using one device-family versus another device-family.

**Comment by daniel-luberda** [comment]
Yes, I agree. It's a bit "hacky" :)

For now, non-locking transformations are only 2-3s ahead of a locking ones, with 200 images x 5 transformations. I think the pros (more balanced memory / cpu usage) are more important, so we can keep locking for a better mobile devices stability.

**Comment by daniel-luberda** [comment]
There's also an old branch with a better performing transformations: https://github.com/luberda-molinet/FFImageLoading/tree/windows-new-bitmapholder-optimization - it may have some significant improvements on Windows. (UWP changed `WriteableBitmap` underlying pixel data from int array to byte array, FFImageLoading is using int array (caveat from Silverlight). But it will force users to refactor all existing custom transformations)
